### PR TITLE
Enable changing of permissions for log directory

### PR DESCRIPTION
**Pull request by itwasntandy**
Previously the permissions for `node['chef_client']['log_dir']` were
hardcoded to 00750.
Sometimes it can be helpful for non-root users to view Chef logs,
and at othertimes, users may wish to set log_dir to /var/log.

In both cases it can be useful to allow the easy changing of mode
for this directory. This change permits it via an attribute.

### CODE REVIEW

**Comment by robcoward** [comment]
:+1:

**Comment by jtimberman** [comment]
I'm still :-1: on this idea, as I was here https://github.com/opscode-cookbooks/chef-client/pull/257

Creating attributes for this kind of thing, to make every single bit customizable defeats the purpose of a general cookbook that fits the broadest general use case. It also increases the complexity of maintenance and testing because we have two conditions to test against.

**Comment by itwasntandy** [comment]
I take your point with regard to complexity and maintenance. However equally burying hardcoded special case defaults cause pain and disruption..

If the goal was to be generic, and simple, why have the special case for the log_dir directory, in setting the mode to 0750 ?
``` 
mode 00750 if dir == 'log_dir'
```
Why not just rely on what has been configured for OS defaults in that case? - as that would truly leave the end result up to the user.

Hardcoding permissions like this can be disruptive, as it's somewhat non obvious to a user of such a cookbook.

We hit this because we wanted to enable log rotation for Chef logs, and run chef from cron.
Previously `node['chef_client']['cron']['log_file]` was set to /var/log/chef-client.log

The log rotation stuff in the chef client cookbook relies on the values of  `node['chef_client']['log_dir']` and `node['chef_client']['log_file'] so we set those appropriately (i.e. log_dir to /var/log, and log_file to chef_client.log) and altered the cron log_file to be a File.join of those attributes.

The net result was /var/log being set to 00750, which was somewhat disruptive.

Of course _now_ we have test cases in place of the permissions of /var/log but the lack of documented clarity regarding the effect of setting `node['chef_client']['log_dir']` to an existing directory caused avoidable disruption.

**Comment by jtimberman** [comment]
> If the goal was to be generic, and simple, why have the special case for the log_dir directory, in setting the mode to 0750 ?

That particular line has a fairly sordid history which took some digging to find. Logic around the log directory mode specifically was added in dde507750523ee86332b45edad23b6461a810d57 which is where it appears the permissions were changed to remove "other" read/execute permission per [COOK-2052](https://tickets.opscode.com/browse/COOK-2052) as an "insecure default." If anything, I'd argue against the log dir original mode of `00755` being an insecure default, and *that* it should be the default hard coded value, effectively reverting that change.

My point about being simple is not about what we've done in the past, but how we maintain this in the future. This cookbook is very, clearly, absolutely, **not** generic and simple. It has cruft and technical debt that makes it, in my opinion, difficult to maintain going forward. It tries to do too much and provide too many knobs for the user to customize because it accommodates so many use cases and bikesheds. Adding more attributes increases complexity. More attributes means more testing to guard against regressions.

> The log rotation stuff in the chef client cookbook relies on the values of node['chef_client']['log_dir'] and node['chef_client']['log_file'] so we set those appropriately (i.e. log_dir to '/var/log', and log_file to 'chef_client.log') and altered the cron log_file to be a File.join of those attributes.

IMO we shouldn't have a `log_dir` attribute. We should have only `log_file` and have a helper method that can return the log directory using `File.dirname`. At the very least, the README should be clear that this directory should be specifically for Chef, and not a general purpose directory. That's a common convention with non-system daemons, e.g. `/var/log/apache2` or similar.

To summarize the above, rather than adding another attribute, I prefer that we:

1. Clarify the documentation about the `node['chef_client']['log_dir']` attribute that it is for Chef's logs, and should not be set to a shared log location.
2. Set the mode to `00755` as it was before commit dde507750523ee86332b45edad23b6461a810d57.

**Comment by itwasntandy** [comment]
> To summarize the above, rather than adding another attribute, I prefer that we:

> 1. Clarify the documentation about the `node['chef_client']['log_dir']` attribute that it is for Chef's logs, and >     should not be set to a shared log location.
> 2. Set the mode to `00755` as it was before commit [dde5077](https://github.com/opscode-cookbooks/chef-client/commit/dde507750523ee86332b45edad23b6461a810d57).

+1 on that. That sounds like a good approach. Would you like me to prepare a pull request for that? - or shall I update this one with that?

Thank you also for explaining a little on the future direction of the cookbook. I've noticed various inconsistencies in there, and will be happy to help tidy and simplify it in future.

**Comment by jtimberman** [comment]
Make a new PR.

**Comment by jtimberman** [comment]
And thanks :cake:

**Comment by chef-supermarket** [comment]
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.chef.io/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## Non-GitHub Verified Committers

There are 1 commit author(s) whose commits are authored by a non-GitHub verified email address. Chef will have to manually verify that they are authorized to contribute.

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

**Comment by juliandunn** [comment]
We've settled on just changing the log directory's permissions `0755` in #257 so I'm closing this now.
